driver-core/src/main/com/mongodb/internal/connection/DnsMultiServerCluster.java

```diff
@@ -58,13 +60,27 @@ public void initialize(final Collection<ServerAddress> hosts) {
             }
 
             private Collection<ServerAddress> applySrvMaxHosts(final Collection<ServerAddress> hosts) {
-                Collection<ServerAddress> newHosts = hosts;
+                Collection<ServerAddress> newHosts = new ArrayList<>(hosts);
+                List<ServerAddress> existingHostList = DnsMultiServerCluster.this.getCurrentDescription().getServerDescriptions()
+                        .stream().map(ServerDescription::getAddress).collect(Collectors.toList());
+                Integer numCurrentEndPoints = existingHostList.size();
+                Integer numRemovedEndPoints = 0;
+                Iterator<ServerAddress> iterator = existingHostList.iterator();
+                while (iterator.hasNext()) {
+                    ServerAddress host = iterator.next();
+                    if (!hosts.contains(host)) {
+                        iterator.remove();
+                        numRemovedEndPoints++;
+                    }
+                }
                 Integer srvMaxHosts = getSettings().getSrvMaxHosts();
                 if (srvMaxHosts != null && srvMaxHosts > 0) {
                     if (srvMaxHosts < hosts.size()) {
                         List<ServerAddress> newHostsList = new ArrayList<>(hosts);
+                        newHostsList.removeAll(existingHostList);
```

Comment:
I am not sure that this is correct. The first step according to the Jira is:

"Remove all nodes no longer present in the SRV response"

The existing host list is being removed here. This is no longer the existing host list, but rather the existing (prior) host list, minus "dead" hosts. But we are removing it, rather than removing the dead hosts.

Perhaps I am misreading? If I am not, then shouldn't the test catch this?

Reply:
This is fine, it's being removed from the new hosts list, which will then be shuffled and added to the results.